suite/tests/linux/syscall_pwait.c

```diff
@@ -0,0 +1,157 @@
+#include <sys/epoll.h>
+#include <unistd.h>
+#include <time.h>
+#include <poll.h>
+
+static void
+signal_handler(int sig, siginfo_t *siginfo, void *context)
+{
+    print("signal received: %d\n", sig);
+}
+
+int
+main(int argc, char *argv[])
+{
+    struct sigaction act;
+    sigset_t new_set;
+
+    INIT();
+
+    intercept_signal(SIGUSR1, (handler_3_t)signal_handler, true);
+    intercept_signal(SIGUSR2, (handler_3_t)signal_handler, true);
+    print("handlers for signals: %d, %d\n", SIGUSR1, SIGUSR2);
+    sigemptyset(&new_set);
+    sigaddset(&new_set, SIGUSR1);
+    sigprocmask(SIG_BLOCK, &new_set, NULL);
+    print("signal blocked: %d\n", SIGUSR1);
+
+    sigset_t test_set;
+    sigfillset(&test_set);
+    sigdelset(&test_set, SIGUSR1);
+    sigdelset(&test_set, SIGUSR2);
+
+    struct timespec sleeptime;
+    sleeptime.tv_sec = 0;
+    sleeptime.tv_nsec = 500 * 1000 * 1000;
+
+    int pid = fork();
+    if (pid < 0) {
+        perror("fork error");
+    } else if (pid == 0) {
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR2);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        return 0;
+    }
+
+    int epoll_fd = epoll_create1(02000000);
+    struct epoll_event events;
+    int count = 0;
+    while (count++ < 3) {
+        /* XXX i#3240: DR currently does not handle the atomicity aspect of this system
+         * call. Once it does, please include this in this test or add a new test. */
+        epoll_pwait(epoll_fd, &events, 24, -1, &test_set);
+    }
+
+    /* waste some time */
+    nanosleep(&sleeptime, NULL);
+
+    pid = fork();
+    if (pid < 0) {
+        perror("fork error");
+    } else if (pid == 0) {
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR2);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        return 0;
+    }
+
+    count = 0;
+    while (count++ < 3) {
+        /* XXX i#3240: DR currently does not handle the atomicity aspect of this system
+         * call. Once it does, please include this in this test or add a new test. */
+        pselect(0, NULL, NULL, NULL, NULL, &test_set);
+    }
+
+    /* waste some time */
+    nanosleep(&sleeptime, NULL);
+
+    pid = fork();
+    if (pid < 0) {
+        perror("fork error");
+    } else if (pid == 0) {
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR2);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        /* waste some time */
+        nanosleep(&sleeptime, NULL);
+        kill(getppid(), SIGUSR1);
+        return 0;
+    }
```

Comment:
To avoid duplicating this code sequence 3 times, what do you think about putting it into a helper function.